rsa/genrsa.c

```diff
@@ -0,0 +1,124 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <openssl/rsa.h>
+#include <openssl/evp.h>
+#include <openssl/bn.h>
+#include <openssl/err.h>
+#include <openssl/pem.h>
+
+char *prtkeyfile=NULL;
+BIO  *bio_log=NULL;
+unsigned int rsabits = 2048;
+
+void usage(void){
+	printf("\n rsagen -s <rsa size in bits> -o <private key file>");
+	printf("\n rsagen -h  -- For help");
+	printf("\n");
+}
+	
+
+void generate_rsa(void){
+	BIGNUM *bn=NULL;
+ 	RSA *rsa=NULL;
+	EVP_PKEY *pkey=NULL;
+	BIO *prtbio;
+
+	if((bn = BN_new()) == NULL) {
+		BIO_printf(bio_log, "Failed to allocate BN\n");
+		goto end;
+	} 
+	BN_set_word(bn, RSA_F4);
+
+	rsa = RSA_new();	
+	RSA_generate_key_ex(rsa, rsabits, bn, NULL);
+
+	if((pkey = EVP_PKEY_new()) == NULL) {
+		BIO_printf(bio_log, "EVP_PKEY_New failed\n");
+		goto end;
+	}
+	EVP_PKEY_assign_RSA(pkey, rsa);
+
+	if (prtkeyfile) {
+		if ((prtbio = BIO_new_file(prtkeyfile, "w")) == NULL) {
+			BIO_printf(bio_log, "\n BIO_new_file %s failed\n", prtkeyfile);
+			goto end;
+		}
+	}
+	else {
+		if ((prtbio = BIO_new_fp(stdout, BIO_NOCLOSE)) == NULL) {
+			BIO_printf(bio_log, "\n BIO_new_file  failed\n");
+			goto end;
+		}
+	}
+
+	if (!PEM_write_bio_PrivateKey(prtbio, pkey, NULL, NULL, 0, NULL, NULL)) {
```

Comment:
Output pem can be RSAPrivateKey(pkcs1?) or PrivateKey(pkcs8). The above one writes the key in pkcs8, what is our intention here whether to have pkcs1 or pkcs8, do you want to take an option from the user?.